### DISCUSSION TITLE
Time Series | Tuning Custom Models & Pipelines

### DISCUSSION THREAD

**Discussion by ngupta23** [pycaret/pycaret#1954] [FAQs]
# Suggested Previous Reads

   - **https://github.com/pycaret/pycaret/discussions/1867**
   - **https://github.com/pycaret/pycaret/discussions/1791**
   - **https://github.com/pycaret/pycaret/discussions/1795**

# Resources for this FAQ

- **[Jupyter Notebook](https://nbviewer.org/github/ngupta23/pycaret_faqs/blob/main/time_series/pycaret_ts_custom_model_sktime_tuning.ipynb)** (can be opened in Google Colab)

# Overview

In the previous FAQ (linked above), we saw how we can train a custom timeseries pipeline using `pycaret`. The only pre-requisite is that this pipeline must be `sktime` compatible. In this FAQ, we will see how we can optimize the hyperparameters for a custom pipeline using the `tune_model()` functionality in `pycaret`. Let's see how.

# Solution

For the purpose of this illustration, we will use the famous airline dataset. We will first model it using an inbuilt model - the `ThetaForecaster`. You can find more information about this forecaster **[here](https://www.sktime.org/en/v0.9.0/api_reference/auto_generated/sktime.forecasting.theta.ThetaForecaster.html#sktime.forecasting.theta.ThetaForecaster)**. Then, we will then try to improve the model using a custom pipeline that includes preprocessing. The forecaster in the custom pipeline will remain the same i.e. `ThetaForecaster`, so that we can compare with the original model. 

```python

#### Get Data ----
y = get_data("airline", verbose=False)

#### Setup Experiment ----
exp = TimeSeriesExperiment()
exp.setup(data=y, fh=12, session_id=42)
```

### Create Model

```python
model = exp.create_model("theta")
```
![image](https://user-images.githubusercontent.com/33585645/146055704-4f2b9bcd-f5f6-4b14-9da0-71008001796b.png)

The baseline model has a large error across the cross-validation folds, e.g. mean MAE is ~ 28. 

```python
exp.predict_model(model)
```

![image](https://user-images.githubusercontent.com/33585645/146055886-2128afe8-9a3d-4a44-92b7-227e549a3310.png)

The score is slightly better across the test set but still high.

```python
exp.plot_model(model)
```

![image](https://user-images.githubusercontent.com/33585645/146056001-f44d11f2-4716-4906-8930-bf67d0ba2b59.png)

The plot also shows that the forecast is underpredicting the values. Let's see if we can tune this model to get better results.

### Tune Model

```python
tuned_model = exp.tune_model(model)
```

![image](https://user-images.githubusercontent.com/33585645/146056193-72955166-66b5-46a9-a09d-490cdcdb372a.png)

There is a slight improvement in the CV results over the earlier model, e.g. MAE is ~ 25 now.

```python
exp.predict_model(tuned_model)
```

![image](https://user-images.githubusercontent.com/33585645/146056350-8d928265-57d4-4442-99b7-713bd3922d91.png)

However, the test results are worse indicating some overfitting.

Next, let's see if we can improve on this model using a custom pipeline using Box-Cox Transformation.

## Understanding Box-Cox Transformation

Box-Cox transformation is a power transformation that is used to make data more normally distributed and stabilize its variance based on the hyperparameter lambda. You can read more about it **[here](https://www.statisticshowto.com/box-cox-transformation/)** and the implementation in `sktime` can be found **[here](https://www.sktime.org/en/v0.9.0/api_reference/auto_generated/sktime.transformations.series.boxcox.BoxCoxTransformer.html)**. The airline dataset shows hints of increasing variance over time so it may benefit from this transformation. Let's see if we can improve the model's performance using this transformation. We will need to define a custom pipeline to achieve this in `pycaret`.

### Define your custom pipeline!

```python
#### Create a sktime pipeline with preprocessing ----
my_forecaster = TransformedTargetForecaster(
    [
        ("bc", BoxCoxTransformer(sp=12)),
        ("model", ThetaForecaster(deseasonalize=True, sp=12)),
    ]
)
```

### Create Custom Pipeline Model

Next, let's model this custom pipeline with `pycaret`. Modeling is done via the same `create_model` call that is used for native `pycaret` models.

```python
custom_model = exp.create_model(my_forecaster)
```

![image](https://user-images.githubusercontent.com/33585645/146058518-033279c8-4903-45d5-af8d-41ddfcb90e18.png)

Right off the bat, we see that the cross-validation results are better than before, e.g. MAE is ~ 22.

```python
exp.predict_model(custom_model)
```
![image](https://user-images.githubusercontent.com/33585645/146058838-5be95aee-a79e-4de8-9438-1bb767a61e75.png)

The test results also show much improvement and are less than the cross-validation scores so indicative of lack of overfitting.

### Tune Custom Pipeline Model

Tuning of a custom model requires the user passing a custom grid to the `tune_model` method/function. Below, we pass a custom grid with 3 different search spaces for the Box-Cox parameter `lambda` search space.

```python
my_grid = {"bc__bounds": [(-3, 0), (0, 1), (1, 3)]}
tuned_custom_model, tuner = exp.tune_model(custom_model, custom_grid=my_grid, return_tuner=True)
```

![image](https://user-images.githubusercontent.com/33585645/146059086-d0e343cb-7511-4173-8fa7-4f677a398205.png)

The results show a slight improvement over the default custom model.

```python
# Optionally, view the results of the grid search.
# pd.DataFrame(tuner.cv_results_)
```

```python
exp.predict_model(tuned_custom_model)
```
![image](https://user-images.githubusercontent.com/33585645/146060957-02647422-7324-4245-9e14-c3f6d5a86514.png)

This slight improvement can be noticed in the test results also.

```python
exp.plot_model(tuned_custom_model)
```
![image](https://user-images.githubusercontent.com/33585645/146061104-aec79235-e168-4a97-869d-5b2415c90d81.png)

The forecast plot also shows that the predictions are very close to the actual test data unlike earlier when they were underpredicting.

So there we have it. We have been able to improve the performance of the original baseline model using a **custom tuned pipeline**.

# Conclusion

While pycaret provides inbuilt models/pipelines for time series analysis, users are by no means limited to using just those models/pipelines. The flexibility to add user defined models/pipelines and **tune these pipelines** is what makes it `pycaret` really shine. So next time you do not find a model/pipeline in pycaret, try adding and tuning your own custom one. 

That's it for now. Happy forecasting!

**Comment by moezali1** [comment]
PyCaret 4.0 sunset the legacy time-series module (sktime/prophet integration), so this FAQ no longer maps to a supported code path.

If you still need time-series forecasting, use sktime / prophet / statsforecast directly — those projects have surpassed what the PyCaret wrapper offered.

For 4.0 docs, see https://pycaret.org/docs and the cleanup notice at https://github.com/pycaret/pycaret/issues/4177.

Closing as outdated.
